### PR TITLE
Add support for registering fallback endpoints.

### PR DESCRIPTION
**Pull request by abhinav**
Expose request, timeout, and trace in the Request object inside handlers.

### CODE REVIEW

**Comment by abhinav** [comment]
@blampe @breerly @junchaowu

File: docs/faq.rst [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+``BadRequestError`` to the caller. This behavior may be overriden by
+registering a raw endpoint with ``TChannel.FALLBACK``.
+
+.. code-block:: python
+
+    from tchannel import TChannel
+
+    server = TChannel(name='myservice')
+
+    @server.register('raw', TChannel.FALLBACK)
```
Comment (blampe) [inline docs/faq.rst]: Does a fallback need to registered for every scheme?

File: tchannel/request.py [hunk 2]
```diff
@@ -47,6 +47,16 @@ class Request(object):
         The most useful piece of information here is probably
         ``request.transport.caller_name``, which is the identity of the
         application that created this request.
+
+    :ivar service:
+        Name of the service being called.
```
Comment (blampe) [inline tchannel/request.py]: How would this differ from the name you advertised as? Is this related to rd transport header?

File: tchannel/request.py [hunk 3]
```diff
@@ -47,6 +47,16 @@ class Request(object):
...
+
+    :ivar timeout:
+        Amount of time (in seconds) within which this request is expected to
+        finish.
+
+    :ivar trace:
```
Comment (blampe) [inline tchannel/request.py]: Doesn't this already live on the request context? Seems more extensible to attach the whole request context object to the request.

File: tchannel/request.py [hunk 4]
```diff
@@ -54,8 +64,11 @@ class Request(object):
     __slots__ = (
         'body',
         'headers',
+        'service',
```
Comment (junchaowu) [inline tchannel/request.py]: is this just for migration information from old request to new request.

File: docs/faq.rst (hunk 1 above)
Comment (abhinav) [inline docs/faq.rst]: No. Raw counts as all schemes. Basically, the scheme equality check is skipped if the handler is a fallback.

File: tchannel/request.py (hunk 2 above)
Comment (abhinav) [inline tchannel/request.py]: This is the name of the target service that the caller provided. Most of the time it will be "this" service but when you're a forwarding service, it'll be whatever you're supposed to forward this to.

File: tests/test_tchannel.py [hunk 5]
```diff
@@ -376,3 +376,99 @@ def test_listen_duplicate_ports():
...
+        response = yield proxy_server_client.call(
+            scheme=request.transport.scheme,
+            service=request.service,
+            arg1=request.endpoint,
+            arg2=request.headers,
+            arg3=request.body,
+            timeout=request.timeout / 2,
+            retry_on=request.transport.retry_flags,
+            retry_limit=0,
+            shard_key=request.transport.shard_key,
```
Comment (blampe) [inline tests/test_tchannel.py]: rd instead of sk?

File: tchannel/request.py (hunk 2 above)
Comment (blampe) [inline tchannel/request.py]: Ah yeah, maybe make a note that it's most relevant to proxies.

File: tchannel/request.py (hunk 3 above)
Comment (abhinav) [inline tchannel/request.py]: Yeah, it does. Currently, using this will override the context's trace info and use the one from the original request.

You think we should just let the default behavior do its thing? That definitely seems more correct because that way the forwarding service itself will show in the trace too.

Comment (abhinav) [inline tchannel/request.py]: Fixed. Tracing information is not exposed anymore.

File: docs/faq.rst (hunk 1 above)
Comment (abhinav) [inline docs/faq.rst]: This does seem a bit janky. I could add some hacks in place to say scheme isn't required if you're using fallback but that'll complicate an already complicated API.

File: tchannel/request.py (hunk 2 above)
Comment (abhinav) [inline tchannel/request.py]: Added note

File: tchannel/request.py (hunk 4 above)
Comment (abhinav) [inline tchannel/request.py]: No. This is the service name the caller gave in the call. If the service is a proxy that will just forward requests, then the service name of the target service will be here.

File: tests/test_tchannel.py (hunk 5 above)
Comment (abhinav) [inline tests/test_tchannel.py]: I don't understand

File: docs/faq.rst (hunk 1 above)
Comment (abhinav) [inline docs/faq.rst]: Actually I figured out a less dirty way to make this happen. Turns out that the old fallback registration code already expected scheme to be absent. Fix incoming.

File: tests/test_tchannel.py (hunk 5 above)
Comment (abhinav) [inline tests/test_tchannel.py]: Done #325.

Comment (abhinav) [inline tests/test_tchannel.py]: Okay, so `rd` is going to 0.21. This is going to master. I'm gonna hold off on adding it here.

**Comment by blampe** [comment]
LGTM after you fix transport header.

File: tests/test_tchannel.py (hunk 5 above)
Comment (abhinav) [inline tests/test_tchannel.py]: nvm, added to master. #326. Will update this PR once that lands.

File: CHANGES.rst [hunk 6]
```diff
@@ -4,7 +4,11 @@ Changes by Version
 0.20.3 (unreleased)
```
Comment (breerly) [inline CHANGES.rst]: Seems like a minor release yeah?

Comment (abhinav) [inline CHANGES.rst]: true. we keep bumping that other release :(

File: tchannel/request.py (hunk 4 above)
Comment (breerly) [inline tchannel/request.py]: Wonder if `caller` is a better name. @blampe whatchutink?

Comment (breerly) [inline tchannel/request.py]: Aha - nm.

**Comment by breerly** [comment]
:shipit:
